### PR TITLE
kv/client: fix ErrNoPendingRegion because of shared map operation (#1569)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #1569 to release-4.0
You can switch your code base to this Pull Request by using [git-extras](https://github.com/tj/git-extras):
```bash
# In ticdc repo:
git pr https://github.com/pingcap/ticdc/pull/1579
```

After apply modifications, you can push your change to this PR via:
```bash
git push git@github.com:ti-srebot/ticdc.git pr/1579:release-4.0-683ae954d690
```

---

### What problem does this PR solve? 

Fix ErrNoPendingRegion found in test

This error is caused by the following scenario

1. Two regions, naming region-1 and region-2, they belong to the same TiKV store
2. Region-1 is handled, and creates a new TiKV stream
3. Region-2 is dispatched, and in old logic, it will be inserted into `pendingRegions` first.
4. `stream.Recv` meets error, and the whole regions belonged to this store will be re-established. The pending region will be cleaned up, region-2 included
5. The step-3 continues, finds no TiKV stream, which has been deleted in step-4, it creates a new one and establishes request for region-2
6. New event of region-2 is received, but region-2 is no longer in `pendingRegions`

The root cause is the map of `pending regions` are shared in two goroutines and the access boundary is not defined correctly

### What is changed and how it works?

When a new TiKV stream is created, create a new `pendingRegions`, and the old `pendingRegions` is kept in old `receiveFromStream` routine.

### Check List 

Tests 

 - Unit test
 - Integration test

### Release note

- Fix replication may be interrupted because of region operation conflict in kv client

### CODE REVIEW

**Comment by amyangfei** [comment]
This PR is blocked by #1534, I will resolve this PR after #1534 is merged

**Review by amyangfei — APPROVED** [review]

**Review by leoppro — APPROVED** [review]

**Pull request merged**
